dev.c

```diff
@@ -20,67 +20,76 @@
 #include <pci.h>
 #include <dev.h>
 
-u32 dev_locate(void)
+u32 dev_locate(u8 cpu_node)
 {
        return pci_locate(DEV_PCI_BUS,
-                         PCI_DEVFN(DEV_PCI_DEVICE, DEV_PCI_FUNCTION));
+                         PCI_DEVFN(DEV_PCI_DEVICE + cpu_node, DEV_PCI_FUNCTION));
 }
 
-u32 dev_read(u32 dev_cap, u32 function, u32 index)
+u32 dev_read(u8 cpu_node, u32 dev_cap, u32 function, u32 index)
 {
     u32 value;
 
     pci_write(0, DEV_PCI_BUS,
-              PCI_DEVFN(DEV_PCI_DEVICE, DEV_PCI_FUNCTION),
+              PCI_DEVFN(DEV_PCI_DEVICE + cpu_node, DEV_PCI_FUNCTION),
               dev_cap + DEV_OP_OFFSET,
               4,
               (u32)(((function & 0xff) << 8) + (index & 0xff)));
 
     pci_read(0, DEV_PCI_BUS,
-             PCI_DEVFN(DEV_PCI_DEVICE, DEV_PCI_FUNCTION),
+             PCI_DEVFN(DEV_PCI_DEVICE + cpu_node, DEV_PCI_FUNCTION),
              dev_cap + DEV_DATA_OFFSET,
              4, &value);
 
        return value;
 }
 
-void dev_write(u32 dev, u32 function, u32 index, u32 value)
+void dev_write(u8 cpu_node, u32 dev, u32 function, u32 index, u32 value)
 {
     pci_write(0, DEV_PCI_BUS,
-        PCI_DEVFN(DEV_PCI_DEVICE, DEV_PCI_FUNCTION),
+        PCI_DEVFN(DEV_PCI_DEVICE + cpu_node, DEV_PCI_FUNCTION),
         dev + DEV_OP_OFFSET,
         4,
         (u32)(((function & 0xff) << 8) + (index & 0xff)) );
 
     pci_write(0, DEV_PCI_BUS,
-        PCI_DEVFN(DEV_PCI_DEVICE, DEV_PCI_FUNCTION),
+        PCI_DEVFN(DEV_PCI_DEVICE + cpu_node, DEV_PCI_FUNCTION),
         dev + DEV_DATA_OFFSET,
         4, value);
 }
 
-void dev_disable_sl(u32 dev)
+void dev_disable_sl(u8 cpu_node, u32 dev)
 {
-    u32 dev_cr = dev_read(dev, DEV_CR, 0);
-    dev_write(dev, DEV_CR, 0, dev_cr & ~(DEV_CR_SL_DEV_EN_MASK));
+    u32 dev_cr = dev_read(cpu_node, dev, DEV_CR, 0);
+    dev_write(cpu_node, dev, DEV_CR, 0, dev_cr & ~(DEV_CR_SL_DEV_EN_MASK));
 }
 
 void disable_memory_protection(void)
 {
-       u32 dev_cap, sldev;
-
-       dev_cap = dev_locate();
-       if (dev_cap) {
-               /* Older families with remains of DEV */
-               dev_disable_sl(dev_cap);
-               return;
-       }
-
-       /* Fam 17h uses different DMA protection control register */
-       pci_read(0, MCH_PCI_BUS,
-                PCI_DEVFN(MCH_PCI_DEVICE, MCH_PCI_FUNCTION),
-                MEMPROT_CR, 4, &sldev);
-       pci_write(0, MCH_PCI_BUS,
-                 PCI_DEVFN(MCH_PCI_DEVICE, MCH_PCI_FUNCTION),
-                 MEMPROT_CR, 4, sldev & ~(MEMPROT_EN));
+    u32 dev_cap, sldev, vid_did;
+    u8 cpu_node = 0;
+
+    dev_cap = dev_locate(cpu_node);
+    if (dev_cap) {
+        /* Older families with remains of DEV */
+        do {
+            dev_disable_sl(cpu_node, dev_cap);
+
+            dev_cap = dev_locate(++cpu_node);
+        } while (dev_cap);
```

Comment:
This would loop indefinitely (due to masking in PCI_DEVFN) for platforms with 8 nodes. Not sure if such systems exist, but won't hurt to test for it. Same for Fam 17h below.

```suggestion
        } while (dev_cap && cpu_node < 8);
```

Reply:
Did it a bit differently to avoid invoking `dev_locate()` for `cpu_node == 8`: https://github.com/TrenchBoot/secure-kernel-loader/compare/951f504307833b994a00879e7142e33e43c8ee6c..18f535843b22e784dbbc8e1ec11d8dc099b3d0fe#diff-e66fa06aefc9a07840b46aa3b351a1fc30dd82cd358e5d4be9d1306d02b5fb0dL78